### ISSUE TITLE
Prototype cloud-init based process

### ISSUE THREAD

**Issue by katherine-black** [cloudigrade/cloudigrade#755]
In GitLab by @infinitewarp on Nov 9, 2020, 10:43

## Summary

As a cloudigrade sysop, I want houndigrade's operation to inspect images in isolation (per run) and I want houndigrade to be able to inspect more than one image at a time. This means we likely need to abandon ECS since it's difficult to scale up for our needs.

Idea to prototype: try to use [cloud-init](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/user-data.html#user-data-cloud-init) (also see [these cloud-init docs](https://cloudinit.readthedocs.io/en/latest/topics/datasources/ec2.html)) to define a process by which we start a new EC2 instance for each volume and provide an initial "kickstart" script that will automatically begin inspection of the attached volume and report results to SQS similar to the current process.

Expected behavior of this prototype:

- create a standalone script that just uses python and boto3 to do these things:
  - talk to AWS to create an EC2 instance with the Volume attached and an init script that does *something*
  - we could use a pre-built basic AMI for this
  - the commands given to cloud-init will download a script (could be tarball to explode)
  - wait and ensure the script actually did something interesting
  - the called script sends a basic "hello world" into an SQS message to verify we gave it correct configs

Note that this prototype does not integrate with cloudigrade nor does it perform an actual full inspection. We just want to prove that cloud-init can do what we want it to do.

Future steps if successful, not part of this story:

- replace cloudigrade's ECS process with this
- teach cloudigrade to do this for each AMI/volume, not batching
- teach cloudigrade or reaper to check the health/life of these new EC2 instances to ensure they're not stuck
- give the downloaded houndigrade script a command to `shutdown -h` at the end (maybe with a delay)
- configure the instances as spot instances
- configure the instances to terminate upon shutdown; see [these docs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/terminating-instances.html#Using_ChangingInstanceInitiatedShutdownBehavior)
- figure out what we need to do to capture logs in CloudWatch

## Acceptance Criteria

- [ ] Verify a standalone python script exists to do the above "Expected behavior of this prototype"

## Assumptions and Questions

- No specific QE effort here since it's just a dev prototype.
- The script(s) we write here do not need to be checked in as part of cloudigrade. They can live in snippets or shared elsewhere. It will live more permanently later when we actually build the cloudigrade changes.
- Assuming this works as expected, after we build the actual cloudigrade version of this, that should unblock https://gitlab.com/cloudigrade/houndigrade/-/issues/38.

**Comment by katherine-black** [comment]
In GitLab by @infinitewarp on Nov 9, 2020, 14:28

changed title from **Prototype clou{-t-}-init based process** to **Prototype clou{+d+}-init based process**

**Comment by werwty** [comment]
Here is the standalone prototype that 
1. creates an instance with a cloud-init script to read a file from a volume and sends the contents to a sqs queue
2. attaches the volume
3. reboots the instance so cloud-init script runs on the attached volume

[749.zip](https://github.com/cloudigrade/cloudigrade/files/5893421/749.zip)

**Comment by katherine-black** [comment]
Thank you for the notebook, closing.
